annif/lexical/mllm.py

```diff
@@ -0,0 +1,215 @@
+                                            params['language']):
+                terms.append(Term(subject_id=subj_id,
+                                  label=label,
+                                  is_pref=False))
+
+        return (terms, subject_ids)
+
+    def _prepare_relations(self, graph, vocab):
+        self._broader_matrix = make_relation_matrix(
+            graph, vocab, SKOS.broader)
+        self._narrower_matrix = make_relation_matrix(
+            graph, vocab, SKOS.narrower)
+        self._related_matrix = make_relation_matrix(
+            graph, vocab, SKOS.related)
+        self._collection_matrix = make_collection_matrix(graph, vocab)
+
+    def _prepare_train_index(self, vocab, analyzer, params):
+        graph = vocab.as_graph()
+        terms, subject_ids = self._prepare_terms(graph, vocab, params)
+        self._prepare_relations(graph, vocab)
+
+        self._vectorizer = CountVectorizer(
+            binary=True,
+            tokenizer=analyzer.tokenize_words
+        )
+        label_corpus = self._vectorizer.fit_transform((t.label for t in terms))
+
+        self._index = TokenSetIndex()
+        for term, label_matrix in zip(terms, label_corpus):
+            tokens = label_matrix.nonzero()[1]
+            tset = TokenSet(tokens, term.subject_id, term.is_pref)
+            self._index.add(tset)
+
+        return subject_ids
+
+    def _calculate_idf(self, subject_ids, doc_count):
+        idf = collections.defaultdict(float)
+        for subj_id in subject_ids:
+            idf[subj_id] = math.log((doc_count + 1) /
+                                    (self._doc_freq[subj_id] + 1)) + 1
+
+        return idf
+
+    def prepare_train(self, corpus, vocab, analyzer, params):
+        subject_ids = self._prepare_train_index(vocab, analyzer, params)
+
+        # frequency of subjects (by id) in the generated candidates
+        self._doc_freq = collections.Counter()
+        # frequency of manually assigned subjects ("domain keyphraseness")
+        self._subj_freq = collections.Counter()
+        doc_count = 0
+        train_x = []
+        train_y = []
+        for idx, doc in enumerate(corpus.documents):
+            doc_subject_ids = [vocab.subjects.by_uri(uri)
+                               for uri in doc.uris]
+            self._subj_freq.update(doc_subject_ids)
+            candidates = self.generate_candidates(doc.text, analyzer)
+            self._doc_freq.update([c.subject_id for c in candidates])
+            train_x.append(candidates)
+            train_y += [(c.subject_id in doc_subject_ids) for c in candidates]
+            doc_count += 1
+
+        # precalculate idf values for all candidate subjects
+        self._idf = self._calculate_idf(subject_ids, doc_count)
+
+        return (np.vstack([self._candidates_to_features(candidates)
+                           for candidates in train_x]), np.array(train_y))
+
+    def _create_classifier(self, params):
+        return BaggingClassifier(
```

Comment:
`BaggingClassifier` has `n_jobs` parameter for parallelization but I don't know could it speed much the overall fitting (or predictions), depending on if candidate generation is much slower.

Reply:
You are right. The [default value](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html) for `n_jobs`in BaggingClassifier is None, which in practice means no parallelism. Using -1 instead would use all CPU cores for both `fit` and `predict` operations. My guess is that the impact would be marginal, but I'm not really sure, so maybe I'll try. My hunch is that candidate generation will dominate, but this may be different for hyperparameter optimization at least, since in that case the candidates are generated just once.

Reply:
I tested this using the kirjaesittelyt corpus, on a laptop with 4 cores (8 threads). Results were not very encouraging.

# Wall clock time in minutes:seconds before / after setting n_jobs=-1

* train on maui-train set: 2:36 / 2:40
* eval on test set (non-parallel): 2:37 / 3:10
* eval on test set (jobs=4): 1:19 / 1:27 with warning `UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1`
* hyperopt on validate set (trials=20, jobs=4): 18:33 / 17:52

For train and eval, the processing time actually increased with n_jobs=-1; parallel evaluation caused additional warnings (it's not possible to use multiprocessing on multiple levels). For hyperopt there was a ~4% speedup, but it could have been a measuring error.